### PR TITLE
Update threshold mechanism

### LINKED ISSUE #872: Incorporate Recommendation Threshold Check for Resource Recommendations
Incorporating a threshold check is needed to ensure that recommendations are only provided if they exceed a certain threshold compared to the current resource settings 

Users of kruize can provide their desired threshold value through the input JSON when creating an experiment. The relevant field in the JSON is `recommendation_settings.threshold`. For example, if a user sets the threshold to 10, the recommendation system should only provide recommendations if they exceed the current resource settings by 10%.

Solution:
- Retrieve the "threshold" value from the "recommendation_settings" field in the input JSON (Create Experiment).
- Compare the threshold value with the difference percentage of calculated resource recommendations.
- If the recommendations do not exceed the threshold, do not generate recommendations and provide a message indicating that the current resource settings are already within the desired threshold. (Notification Message: `Workload optimized, no changes needed`)
- If the recommendations exceed the threshold, generate and provide the recommendations as usual.

This enhancement will ensure that users only receive recommendations when there is a significant deviation from their current resource settings, allowing them to optimize their pod configurations effectively.

Please let me know if you have any further suggestions or requirements for this enhancement.

FYI @dinogun

### PR DESCRIPTION
**Pull request by bharathappali**
This PR incorporates the check for threshold and compare with current variation, if the variation is more than threshold it populates the recommendation else notifies that the workload is already optimised 

Fixes #872

### CODE REVIEW

File: src/main/java/com/autotune/analyzer/utils/AnalyzerConstants.java [hunk 1]
```diff
@@ -241,117 +241,137 @@ public int getSeverity() {
     }
 
     public enum RecommendationNotification {
-        DURATION_BASED_RECOMMENDATIONS_AVAILABLE (
+        INFO_DURATION_BASED_RECOMMENDATIONS_AVAILABLE(
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/utils/AnalyzerConstants.java:244]: As I had mentioned previously (in a different PR), we need to move all of these recommendation related constants out of `AnalyzerConstants` and instead inside the `recommendations` package

**Comment by dinogun** [comment]
@AB Looks like the MD changes are missing, please add the appropriate doc

**Comment by bharathappali** [comment]
Use Case - Complete resource optimised (Both CPU and Memory are optimised)
```
{
    "cluster_name": "cluster-one-division-bell",
    "kubernetes_objects": [
        {
            "type": "deployment",
            "name": "my-sample-deployment",
            "namespace": "default",
            "containers": [
                {
                    "container_image_name": "my-sample-image",
                    "container_name": "my-sample-container",
                    "recommendations": {
                        "notifications": {
                            "112101": {
                                "type": "info",
                                "message": "Duration Based Recommendations Available",
                                "code": 112101
                            }
                        },
                        "data": {
                            "2023-07-03T14:57:47.925Z": {
                                "duration_based": {
                                    "short_term": {
                                        "monitoring_start_time": "2023-07-02T14:57:47.925Z",
                                        "monitoring_end_time": "2023-07-03T14:57:47.925Z",
                                        "duration_in_hours": 24.0,
                                        "pods_count": 1,
                                        "confidence_level": 0.0,
                                        "current": {
                                            "limits": {
                                                "memory": {
                                                    "amount": 300.0,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 3.0,
                                                    "format": "cores"
                                                }
                                            },
                                            "requests": {
                                                "memory": {
                                                    "amount": 300.0,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 3.0,
                                                    "format": "cores"
                                                }
                                            }
                                        },
                                        "variation": {
                                            "limits": {
                                                "memory": {
                                                    "amount": 0.0,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 0.0,
                                                    "format": "cores"
                                                }
                                            },
                                            "requests": {
                                                "memory": {
                                                    "amount": 0.0,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 0.0,
                                                    "format": "cores"
                                                }
                                            }
                                        },
                                        "notifications": {
                                            "324004": {
                                                "type": "notice",
                                                "message": "Workload is optimised wrt MEMORY LIMITS, no changes needed",
                                                "code": 324004
                                            },
                                            "324003": {
                                                "type": "notice",
                                                "message": "Workload is optimised wrt MEMORY REQUESTS, no changes needed",
                                                "code": 324003
                                            },
                                            "323004": {
                                                "type": "notice",
                                                "message": "Workload is optimised wrt CPU REQUESTS, no changes needed",
                                                "code": 323004
                                            },
                                            "323005": {
                                                "type": "notice",
                                                "message": "Workload is optimised wrt CPU LIMITS, no changes needed",
                                                "code": 323005
                                            }
                                        }
                                    },
                                    "medium_term": {
                                        "pods_count": 0,
                                        "confidence_level": 0.0,
                                        "notifications": {
                                            "120001": {
                                                "type": "info",
                                                "message": "There is not enough data available to generate a recommendation.",
                                                "code": 120001
                                            }
                                        }
                                    },
                                    "long_term": {
                                        "pods_count": 0,
                                        "confidence_level": 0.0,
                                        "notifications": {
                                            "120001": {
                                                "type": "info",
                                                "message": "There is not enough data available to generate a recommendation.",
                                                "code": 120001
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            ]
        }
    ],
    "version": "1.0",
    "experiment_name": "fascinating-bhanvi"
}
```

**Comment by bharathappali** [comment]
Use Case: CPU optimised

```
{
    "cluster_name": "cluster-one-division-bell",
    "kubernetes_objects": [
        {
            "type": "deployment",
            "name": "my-sample-deployment",
            "namespace": "default",
            "containers": [
                {
                    "container_image_name": "my-sample-image",
                    "container_name": "my-sample-container",
                    "recommendations": {
                        "notifications": {
                            "112101": {
                                "type": "info",
                                "message": "Duration Based Recommendations Available",
                                "code": 112101
                            }
                        },
                        "data": {
                            "2023-07-03T14:50:32.529Z": {
                                "duration_based": {
                                    "short_term": {
                                        "monitoring_start_time": "2023-07-02T14:50:32.529Z",
                                        "monitoring_end_time": "2023-07-03T14:50:32.529Z",
                                        "duration_in_hours": 24.0,
                                        "pods_count": 1,
                                        "confidence_level": 0.0,
                                        "current": {
                                            "limits": {
                                                "memory": {
                                                    "amount": 427.73,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 3.0,
                                                    "format": "cores"
                                                }
                                            },
                                            "requests": {
                                                "memory": {
                                                    "amount": 560.93,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 3.0,
                                                    "format": "cores"
                                                }
                                            }
                                        },
                                        "config": {
                                            "limits": {
                                                "memory": {
                                                    "amount": 1193.076,
                                                    "format": "MiB"
                                                }
                                            },
                                            "requests": {
                                                "memory": {
                                                    "amount": 1193.076,
                                                    "format": "MiB"
                                                }
                                            }
                                        },
                                        "variation": {
                                            "limits": {
                                                "memory": {
                                                    "amount": 765.346,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 0.0,
                                                    "format": "cores"
                                                }
                                            },
                                            "requests": {
                                                "memory": {
                                                    "amount": 632.1460000000001,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 0.0,
                                                    "format": "cores"
                                                }
                                            }
                                        },
                                        "notifications": {
                                            "323004": {
                                                "type": "notice",
                                                "message": "Workload is optimised wrt CPU REQUESTS, no changes needed",
                                                "code": 323004
                                            },
                                            "323005": {
                                                "type": "notice",
                                                "message": "Workload is optimised wrt CPU LIMITS, no changes needed",
                                                "code": 323005
                                            }
                                        }
                                    },
                                    "medium_term": {
                                        "pods_count": 0,
                                        "confidence_level": 0.0,
                                        "notifications": {
                                            "120001": {
                                                "type": "info",
                                                "message": "There is not enough data available to generate a recommendation.",
                                                "code": 120001
                                            }
                                        }
                                    },
                                    "long_term": {
                                        "pods_count": 0,
                                        "confidence_level": 0.0,
                                        "notifications": {
                                            "120001": {
                                                "type": "info",
                                                "message": "There is not enough data available to generate a recommendation.",
                                                "code": 120001
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            ]
        }
    ],
    "version": "1.0",
    "experiment_name": "exquisite-dinakar"
}
```

**Comment by bharathappali** [comment]
Use Case: Memory Optimised

```
{
    "cluster_name": "cluster-one-division-bell",
    "kubernetes_objects": [
        {
            "type": "deployment",
            "name": "my-sample-deployment",
            "namespace": "default",
            "containers": [
                {
                    "container_image_name": "my-sample-image",
                    "container_name": "my-sample-container",
                    "recommendations": {
                        "notifications": {
                            "112101": {
                                "type": "info",
                                "message": "Duration Based Recommendations Available",
                                "code": 112101
                            }
                        },
                        "data": {
                            "2023-07-03T14:52:30.712Z": {
                                "duration_based": {
                                    "short_term": {
                                        "monitoring_start_time": "2023-07-02T14:52:30.712Z",
                                        "monitoring_end_time": "2023-07-03T14:52:30.712Z",
                                        "duration_in_hours": 24.0,
                                        "pods_count": 1,
                                        "confidence_level": 0.0,
                                        "current": {
                                            "limits": {
                                                "memory": {
                                                    "amount": 300.0,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 2.92,
                                                    "format": "cores"
                                                }
                                            },
                                            "requests": {
                                                "memory": {
                                                    "amount": 300.0,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 3.3,
                                                    "format": "cores"
                                                }
                                            }
                                        },
                                        "config": {
                                            "limits": {
                                                "cpu": {
                                                    "amount": 7.62,
                                                    "format": "cores"
                                                }
                                            },
                                            "requests": {
                                                "cpu": {
                                                    "amount": 7.62,
                                                    "format": "cores"
                                                }
                                            }
                                        },
                                        "variation": {
                                            "limits": {
                                                "memory": {
                                                    "amount": 0.0,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 4.7,
                                                    "format": "cores"
                                                }
                                            },
                                            "requests": {
                                                "memory": {
                                                    "amount": 0.0,
                                                    "format": "MiB"
                                                },
                                                "cpu": {
                                                    "amount": 4.32,
                                                    "format": "cores"
                                                }
                                            }
                                        },
                                        "notifications": {
                                            "324004": {
                                                "type": "notice",
                                                "message": "Workload is optimised wrt MEMORY LIMITS, no changes needed",
                                                "code": 324004
                                            },
                                            "324003": {
                                                "type": "notice",
                                                "message": "Workload is optimised wrt MEMORY REQUESTS, no changes needed",
                                                "code": 324003
                                            }
                                        }
                                    },
                                    "medium_term": {
                                        "pods_count": 0,
                                        "confidence_level": 0.0,
                                        "notifications": {
                                            "120001": {
                                                "type": "info",
                                                "message": "There is not enough data available to generate a recommendation.",
                                                "code": 120001
                                            }
                                        }
                                    },
                                    "long_term": {
                                        "pods_count": 0,
                                        "confidence_level": 0.0,
                                        "notifications": {
                                            "120001": {
                                                "type": "info",
                                                "message": "There is not enough data available to generate a recommendation.",
                                                "code": 120001
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            ]
        }
    ],
    "version": "1.0",
    "experiment_name": "glorious-kusuma"
}
```

**Review by dinogun — APPROVED** [review]
LGTM!

**Pull request merged**